pkg/api/v1/pod/util.go

```diff
@@ -50,6 +50,59 @@ func FindPort(pod *v1.Pod, svcPort *v1.ServicePort) (int, error) {
 	return 0, fmt.Errorf("no suitable port for manifest: %s", pod.UID)
 }
 
+// ContainerType specifies the type of container
+type ContainerType int
+
+const (
+	// DontCheckFeatureGate skips checking feature gate
+	DontCheckFeatureGate = 1 << iota
+	// InitContainer is for init container
+	InitContainer
+	// Container is for container
+	Container
+	// EphemeralContainer is for ephemeral container which checks feature gate
+	EphemeralContainer
+	// AllContainers cover all containers
+	AllContainers = InitContainer | Container | EphemeralContainer
+	// NonEphemeralContainers cover all containers except ephemeral container
+	NonEphemeralContainers = InitContainer | Container
+)
+
+// ContainerVisitorV2 is called with container spec of given container type, and returns true
+// if visiting should continue.
+type ContainerVisitorV2 func(container *v1.Container, containerType ContainerType) (shouldContinue bool)
+
+// VisitContainersV2 invokes the visitor function with a pointer to the container
+// spec of given container type in the given pod spec. If visitor returns false,
+// visiting is short-circuited. VisitContainersV2 returns true if visiting completes,
+// false if visiting was short-circuited.
+func VisitContainersV2(podSpec *v1.PodSpec, containerTypes int, visitor ContainerVisitorV2) bool {
```

Comment:
including DontCheckFeatureGate in the container types bitset is confusing. what about making feature-enablement a specific option, either via a bool param or by making a separate method, e.g.:

```
// VisitEnabledContainersByType visits containers of the specified types, if their relevant features are enabled
func VisitEnabledContainersByType(podSpec *v1.PodSpec, containerTypes int, visitor ContainerVisitorV2) bool

// VisitContainersByType visits all containers of the specified types, regardless of feature enablement
func VisitContainersByType(podSpec *v1.PodSpec, containerTypes int, visitor ContainerVisitorV2) bool
```